tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,439 @@
+                previous_spec['outputs'].add(node_name)
+                current_spec['inputs'].add(node_name)
+                
+                node = node_name_to_node_def[node_name]
+                placeholder = _create_placeholder_node_from_existing_node(node, graph)
+                current_spec['subgraph'].node.append(placeholder)
+                
+    return current_spec
+        
+
+def _partition_one_remote_op_layer(remote_op_names, node_name_to_input_names):
+    """Construct spec for remote ops"""
+    list_of_specs = []
+    for remote_op_name in remote_op_names:
+        spec = {'subgraph': None, 
+                'inputs': set(node_name_to_input_names[remote_op_name]), 
+                'outputs': set([remote_op_name]), 
+                'body_nodes': set([remote_op_name]), 
+                'is_remote_op': True, 
+                'nodes_from_other_layers': None}
+        list_of_specs.append(spec)
+    
+    return list_of_specs
+
+
+def _create_placeholder_node(dtype, shape, name):
+    temp = tf.Graph()
+    with temp.as_default():
+        placeholder = tf.compat.v1.placeholder(dtype=dtype, shape=shape, name=name)
+        return temp.as_graph_def().node[0]      # The first and the only node  
+
+
+def _create_placeholder_node_from_existing_node(node, graph):
+    operation = graph.get_operation_by_name('import/%s' % (node.name))
+    dtype = operation.outputs[0].dtype
+    return _create_placeholder_node(dtype=dtype,
+                                    shape=None,
+                                    name=node.name)
+
+
+def _get_inputs_from_subgraph(subgraph):
+    inputs = set([node.name for node in subgraph.node if _check_placeholder_op(node)])
+    return inputs
+
+    
+def _get_regular_nodes_from_subgraph(subgraph):
+    regular_nodes = set([node.name for node in subgraph.node
+                         if not _check_placeholder_op(node)])
+    return regular_nodes
+
+
+class Relations(object):
+    """A class that outputs remote op layers (custom topological sort).
+    
+    What is a layer? A layer is a set of nodes that are ready to execute."""
+    def __init__(self, relations):
+        self.relations = relations
+        self.processed = set([])
+        self.to_be_processed = set(relations.keys())
+    
+    def check_if_finished(self):
+        return not self.to_be_processed
+    
+    def get_next_layer(self):
+        layer_nodes = set([])
+        
+        for node in self.to_be_processed:
+            node_inputs = set(self.relations[node])
+            if node_inputs.issubset(self.processed):
+                layer_nodes.add(node)
+                
+        for node in layer_nodes:
+            self.to_be_processed.remove(node)
+            self.processed.add(node)
+        
+        return layer_nodes
+    
+    def _debug_print_layers(self):
+        while not self.check_if_finished():
+            print(self.get_next_layer())
+            
+            
+            
```

Comment:
nit: file should end with only one newline

Reply:
Discussed before